### PR TITLE
Allow public_suffix 3

### PR DESCRIPTION
**Pull request by twalpole**
Allow the use of public_suffix 3.x with addressable

### CODE REVIEW

File: tasks/gem.rake [hunk 1]
```diff
@@ -20,7 +20,7 @@ namespace :gem do
 
     s.required_ruby_version = '>= 2.0'
 
-    s.add_runtime_dependency 'public_suffix', '~> 2.0', '>= 2.0.2'
+    s.add_runtime_dependency 'public_suffix', '>= 2.0.2', '< 4.0'
```
Comment (jesperronn) [inline tasks/gem.rake]: @houndci-bot should be consistent to all lines (and not only the affected ones), and is irrelevant for the change here.

(Yes I know you're a bot)

**Comment by jesperronn** [comment]
background info found in https://github.com/weppos/publicsuffix-ruby/blob/master/CHANGELOG.md#release-300 and https://github.com/weppos/publicsuffix-ruby/pull/133 

This was a major release in `public_suffix`, but if its not affecting any usages in this library, then it's probably only a minor release of addressable? @sporkmonger what do you think?

**Comment by dentarg** [comment]
@jesperronn  3.0.0 of `public_suffix` dropped support for Ruby 2.0, but Addressable 
 still support it: https://github.com/sporkmonger/addressable/blob/addressable-2.5.1/.travis.yml#L9.

[Not sure](https://github.com/weppos/publicsuffix-ruby/issues/127) if that should warrant a bigger bump.

**Comment by dentarg** [comment]
Ah, it doesn't matter for this change. Ruby 2.0 will use Public Suffix 2.0.5 (https://travis-ci.org/sporkmonger/addressable/jobs/261162023).

**Comment by sporkmonger** [comment]
If `public_suffix` is following semver and the only breaking change is that they stopped supporting 2.0 then I'm fine w/ taking the approach of allowing either version under the assumption that any new breaking change will result in a 4.0 version.

File: tasks/gem.rake (hunk 1 above)
Comment (sporkmonger) [inline tasks/gem.rake]: Yeah, at some point I'll probably do a big code style audit. No worries for this PR.

**Comment by twalpole** [comment]
@sporkmonger There were other breaking changes around PublicSuffix::List, but not in any methods that addressable uses.

File: gemfiles/Gemfile.public_suffix_2 [hunk 2]
```diff
@@ -0,0 +1,34 @@
+source 'https://rubygems.org'
```
Comment (sporkmonger) [inline gemfiles/Gemfile.public_suffix_2]: Just to clarify, this `Gemfile` is only for testing purposes? If so, would it make more sense to name it in such a way that that's obviously, or perhaps include a comment to that effect?

**Comment by sporkmonger** [comment]
OK, that should be OK. We have decent test coverage, so it ought to be alright.

File: gemfiles/Gemfile.public_suffix_2 (hunk 2 above)
Comment (twalpole) [inline gemfiles/Gemfile.public_suffix_2]: @sporkmonger It's for dev or test, just like the main /Gemfile (which isn't used when building the gem, the gemspec is) which doesn't have comments about that.

**Comment by sporkmonger** [comment]
Alright cool. Might rearrange things a bit later when I have time, but seems fine for now.

**Comment by farwayer** [comment]
Any chance new version will be released soon? ruby-addressable is broken in [archlinux](https://www.archlinux.org/packages/community/any/ruby-addressable/) now because public_suffix [was updated to 3.0](https://www.archlinux.org/packages/community/any/ruby-public_suffix/) in repos 😞

**Comment by sporkmonger** [comment]
Done.

**Comment by farwayer** [comment]
Thank you!

**Comment by ybart** [comment]
Nice! Thanks 😀

**Pull request merged**